tools/chapel-py/scripts/generate-pyi.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python3
+#
+# Copyright 2020-2023 Hewlett Packard Enterprise Development LP
+# Copyright 2004-2019 Cray Inc.
+# Other additional copyright holders may be indicated within.
+#
+# The entirety of this work is licensed under the Apache License,
+# Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License.
+#
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import chapel.core
+from typing import List, Tuple
+
+
+def get_base_header() -> str:
+    c = chapel.core.Context()
+    return c._get_pyi_file()
+
+
+def _wrap_docstring(docstring: str, indent: int = 8) -> List[str]:
+    spaces = " " * indent
+    triple_quote = spaces + '"""'
+    return [triple_quote, spaces + docstring, triple_quote]
+
+
+def _def_body(indent: int = 8) -> List[str]:
+    spaces = " " * indent
+    return [spaces + "..."]
+
+
+def _wrap_method(
+    name: str,
+    args: List[Tuple[str, str | None]] = [],
+    rettype: str = "None",
+    indent: int = 4,
+) -> List[str]:
+    spaces = " " * indent
+    argstr = ""
+    for a, t in args:
+        typestr = f": {t}" if t else ""
+        argstr += f", {a}{typestr}"
+    return [spaces + f"def {name}(self{argstr}) -> {rettype}:"]
+
+
+def get_Context_header() -> str:
+    s = [
+        "class Context:",
+        *_wrap_method("parse", args=[("path", "str")], rettype="List[AstNode]"),
+        *_wrap_docstring("Parse a top-level AST node from the given file"),
+        *_def_body(),
```

Comment:
These three are used over and over again which makes me think you should use a helper function for this logic.